src/Distribution/Extra/Doctest.hs

```diff
@@ -304,6 +330,30 @@ generateBuildModule testSuiteName flags pkg lbi = do
        isSpecific (SpecificPackageDB _) = True
        isSpecific _                     = False
 
+    mbLibraryName :: Library -> Maybe String
+#if MIN_VERSION_Cabal(2,0,0)
+    -- Cabal-2.0 introduced internal libraries, which are named.
+    mbLibraryName = fmap unUnqualComponentName . libName
+#else
+    -- Before that, there was only ever at most one library per
+    -- .cabal file, which has no name.
+    mbLibraryName _ = Nothing
+#endif
+
+    executableName :: Executable -> String
+#if MIN_VERSION_Cabal(2,0,0)
+    executableName = unUnqualComponentName . exeName
+#else
+    executableName = exeName
+#endif
+
+    exeMainIs :: Executable -> [ModuleName]
+    exeMainIs exe = [ ModuleName.fromString
+                    $ dropExtension -- Be careful: the IsString instance doesn't
+                                    -- like "Main.hs", only "Main"
+                    $ modulePath exe
```

Comment:
executable file name may be `Something.hs`, when the module is still `Main`, for it we need to pass actual file name.